### PR TITLE
Add warning if a resolution failed

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
r? @QuietMisdreavus

### CODE REVIEW

**Comment by oli-obk** [comment]
Can you add a ui test showing the new warnings?

**Comment by GuillaumeGomez** [comment]
Good idea!

File: src/bootstrap/test.rs [hunk 1]
```diff
@@ -514,6 +514,44 @@ impl Step for RustdocJS {
...
+    const DEFAULT: bool = true;
+    const ONLY_HOSTS: bool = true;
+
+    fn should_run(run: ShouldRun) -> ShouldRun {
+        run.path("src/test/rustdoc-ui")
+    }
+
+    fn make_run(run: RunConfig) {
+        let compiler = run.builder.compiler(run.builder.top_stage, run.host);
+        run.builder.ensure(tool::Rustdoc {
```
Comment (Mark-Simulacrum) [inline src/bootstrap/test.rs]: No need for this as `builder.rustdoc` below will take care of it for you.

Comment (GuillaumeGomez) [reply src/bootstrap/test.rs]: Where?

**Comment by GuillaumeGomez** [comment]
Updated (I also added some useful options to rustdoc).

File: src/bootstrap/test.rs (hunk 1 above)
Comment (Mark-Simulacrum) [reply src/bootstrap/test.rs]: `cmd.arg("--rustdoc-path").arg(builder.rustdoc(compiler.host));` on what might be 898 -- that call to `rustdoc` will auto-ensure it for you, so you shouldn't need this.

File: src/test/rustdoc-ui/intra-links-warning.stderr [hunk 2]
```diff
@@ -0,0 +1,6 @@
+warning: [src] cannot be resolved, ignoring it...
```
Comment (oli-obk) [inline src/test/rustdoc-ui/intra-links-warning.stderr]: What's with this src thing?

File: src/bootstrap/test.rs [hunk 3]
```diff
@@ -514,6 +514,44 @@ impl Step for RustdocJS {
     }
 }
 
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub struct RustdocUi {
```
Comment (oli-obk) [inline src/bootstrap/test.rs]: This is a lot of code for very few tests. I was hoping we could somehow express this in the regular ui tests by inserting some flags via comments in the test?

File: src/librustdoc/clean/mod.rs [hunk 4]
```diff
@@ -1160,6 +1160,10 @@ enum PathKind {
     Type,
 }
 
+fn resolution_failure(cx: &DocContext, path_str: &str) {
+    cx.sess().warn(&format!("[{}] cannot be resolved, ignoring it...", path_str));
```
Comment (oli-obk) [inline src/librustdoc/clean/mod.rs]: Are there spans available that we can use to report the warnings?

File: src/librustdoc/core.rs [hunk 5]
```diff
@@ -153,14 +156,42 @@ pub fn run_core(search_paths: SearchPaths,
             edition,
             ..config::basic_debugging_options()
         },
+        error_format,
         ..config::basic_options().clone()
     };
 
     let codemap = Lrc::new(codemap::CodeMap::new(sessopts.file_path_mapping()));
-    let diagnostic_handler = errors::Handler::with_tty_emitter(ColorConfig::Auto,
```
Comment (oli-obk) [inline src/librustdoc/core.rs]: What is the purpose of these changes?

File: src/librustdoc/clean/mod.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/mod.rs]: It'll be part of another pr. I think this one is already big enough.

File: src/librustdoc/core.rs (hunk 5 above)
Comment (GuillaumeGomez) [reply src/librustdoc/core.rs]: Allow rustdoc output to be checked through UI test suite.

File: src/test/rustdoc-ui/intra-links-warning.stderr (hunk 2 above)
Comment (GuillaumeGomez) [reply src/test/rustdoc-ui/intra-links-warning.stderr]: That's actually a good question... Looking where it's coming from.

File: src/bootstrap/test.rs (hunk 3 above)
Comment (QuietMisdreavus) [reply src/bootstrap/test.rs]: As would i, but (as far as i know) there's no facility in compiletest to swap out which binary is called for a specific test - only which arguments are passed. In addition, we would have to tell bootstrap to have rustdoc ready for the UI tests, even if someone is trying to run a specific non-rustdoc test.

Plus, we wanted to introduce UI tests for rustdoc at some point anyway. `>_>` Having them in a separate test directory will help us when we add more over time.

Comment (GuillaumeGomez) [reply src/bootstrap/test.rs]: We need to have a struct directly to bind the test folder path.

**Comment by QuietMisdreavus** [comment]
The rustdoc changes look good to me, but i'm less certain about the changes to bootstrap and compiletest. I'd like some extra eyes to look at those bits.

**Comment by GuillaumeGomez** [comment]
cc @oli-obk

File: src/bootstrap/test.rs [hunk 6]
```diff
@@ -514,6 +514,41 @@ impl Step for RustdocJS {
...
```
Comment (Mark-Simulacrum) [inline src/bootstrap/test.rs]: This struct should be replaceable with a `host_test!` macro invocation I believe.

File: src/tools/compiletest/src/main.rs [hunk 7]
```diff
@@ -276,6 +276,8 @@ pub fn parse_config(args: Vec<String>) -> Config {
         ),
     };
 
+    let src_base = opt_path(matches, "src-base");
+    let run_ignored = matches.opt_present("ignored");
```
Comment (Mark-Simulacrum) [inline src/tools/compiletest/src/main.rs]: Avoiding these changes seems good for consistency's sake.

File: src/tools/compiletest/src/runtest.rs [hunk 8]
```diff
@@ -1604,17 +1611,20 @@ impl<'test> TestCx<'test> {
...
 
-        if let Some(ref incremental_dir) = self.props.incremental_dir {
-            rustc.args(&[
-                "-C",
-                &format!("incremental={}", incremental_dir.display()),
-            ]);
-            rustc.args(&["-Z", "incremental-verify-ich"]);
-            rustc.args(&["-Z", "incremental-queries"]);
-        }
+        println!("{:?}", is_rustdoc);
```
Comment (Mark-Simulacrum) [inline src/tools/compiletest/src/runtest.rs]: Leftover debugging code?

File: src/bootstrap/test.rs (hunk 6 above)
Comment (GuillaumeGomez) [reply src/bootstrap/test.rs]: I tried to at first to avoid writing code by hand, but failed... I can't only rely on UI or rustdoc test suite, I need a mix of the two.

File: src/tools/compiletest/src/runtest.rs (hunk 8 above)
Comment (GuillaumeGomez) [reply src/tools/compiletest/src/runtest.rs]: Oops indeed!

File: src/tools/compiletest/src/main.rs (hunk 7 above)
Comment (GuillaumeGomez) [reply src/tools/compiletest/src/main.rs]: How would you do it?

**Review by oli-obk — COMMENTED** [review]
ups, forgot to submit this review...

File: src/test/rustdoc-ui/intra-links-warning.rs [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![deny(warnings)]
+
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/test/rustdoc-ui/intra-links-warning.rs]: It hasn't been implemented in rustdoc (yet?). I'll remove this line.

File: src/bootstrap/test.rs [hunk 10]
```diff
@@ -514,6 +514,41 @@ impl Step for RustdocJS {
...
+    type Output = ();
+    const DEFAULT: bool = true;
+    const ONLY_HOSTS: bool = true;
+
+    fn should_run(run: ShouldRun) -> ShouldRun {
+        run.path("src/test/rustdoc-ui")
+    }
+
+    fn make_run(run: RunConfig) {
+        let compiler = run.builder.compiler(run.builder.top_stage, run.host);
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/bootstrap/test.rs]: Apparently it's done correctly (or at least that's what I understood from @Mark-Simulacrum's comment).

[the remark this answers is not in the document]
Comment (Mark-Simulacrum) [reply src/bootstrap/test.rs]: Yes, this is fine as is, we depend on rustdoc below when we pass it's path.

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by GuillaumeGomez** [comment]
Anyone? cc @oli-obk @Mark-Simulacrum @QuietMisdreavus

File: src/tools/compiletest/src/main.rs (hunk 7 above)
Comment (Mark-Simulacrum) [reply src/tools/compiletest/src/main.rs]: Er, this file should be possible to leave as-is: both changes were simply extracted from struct's definition AFAICT.

**Comment by Mark-Simulacrum** [comment]
r=me on bootstrap changes

**Comment by GuillaumeGomez** [comment]
@Mark-Simulacrum: Any clue why all ui tests seem to have no output anymore? :-/

File: src/bootstrap/test.rs [hunk 11]
```diff
@@ -868,14 +907,22 @@ impl Step for Compiletest {
...
+            vec!["-Crpath".to_string()]
+        };
+        if !is_rustdoc_ui {
+            if build.config.rust_optimize_tests {
+                flags.push("-O".to_string());
+            }
+            if build.config.rust_debuginfo_tests {
+                flags.push("-g".to_string());
+            }
+            flags.push("-Zmiri -Zunstable-options".to_string());
```
Comment (Mark-Simulacrum) [inline src/bootstrap/test.rs]: Bad rebase. This `-Zmiri` no longer exists on Rust's master, so remove it here as well.

**Review by Mark-Simulacrum — CHANGES_REQUESTED** [review]
This should fix empty UI tests.

File: src/bootstrap/test.rs (hunk 11 above)
Comment (GuillaumeGomez) [reply src/bootstrap/test.rs]: Thanks!

**Comment by GuillaumeGomez** [comment]
Arg, new links have been added the docs! ><

**Comment by GuillaumeGomez** [comment]
The appveyor tests output just end on `running 692 tests`. So I have no idea if it's a "normal" failure or not...

**Comment by GuillaumeGomez** [comment]
Rebased, let's go again!

@bors: r+ p=11
